git/ssh_config.go

```diff
@@ -45,44 +46,34 @@ func (m SSHAliasMap) Translator() func(*url.URL) *url.URL {
 	}
 }
 
-// ParseSSHConfig constructs a map of SSH hostname aliases based on user and
-// system configuration files
-func ParseSSHConfig() SSHAliasMap {
-	configFiles := []string{
-		"/etc/ssh_config",
-		"/etc/ssh/ssh_config",
-	}
-	if homedir, err := homedir.Dir(); err == nil {
-		userConfig := filepath.Join(homedir, ".ssh", "config")
-		configFiles = append([]string{userConfig}, configFiles...)
-	}
-
-	openFiles := make([]io.Reader, 0, len(configFiles))
-	for _, file := range configFiles {
-		f, err := os.Open(file)
-		if err != nil {
-			continue
-		}
-		defer f.Close()
-		openFiles = append(openFiles, f)
-	}
-	return sshParse(openFiles...)
+type parser struct {
+	aliasMap SSHAliasMap
 }
 
-func sshParse(r ...io.Reader) SSHAliasMap {
-	config := make(SSHAliasMap)
-	for _, file := range r {
-		_ = sshParseConfig(config, file)
+func (p *parser) read(fileName string) error {
+	file, err := os.Open(fileName)
+	if err != nil {
+		return err
 	}
-	return config
-}
+	defer file.Close()
 
-func sshParseConfig(c SSHAliasMap, file io.Reader) error {
 	hosts := []string{"*"}
 	scanner := bufio.NewScanner(file)
 	for scanner.Scan() {
 		line := scanner.Text()
-		match := sshHostRE.FindStringSubmatch(line)
+		match := sshIncludeDirectiveRE.FindStringSubmatch(line)
+		if match != nil {
+			fileNames, err := filepath.Glob(match[2])
```

Comment:
Interested in how relative paths are supposed to be resolved, I've just looked up the documentation for the `Include` directive and it says:

> Include the specified configuration file(s).  Multiple pathnames may be specified and each pathname may contain glob(7) wildcards and, for user configurations, shell-like `~` references to user home directories.  Files without absolute paths are assumed to be in `~/.ssh` if included in a user configuration file or `/etc/ssh` if included from the system configuration file.  Include directive may appear inside a `Match` or `Host` block to perform conditional inclusion.

So for this `Include` implementation to work properly, I think that is missing is:
- relative path support;
- multiple paths support (I'm guessing paths are split by spaces?);
- the contents of an `Include` following a `Match` or `Host` block should only ever apply to that `Match`/`Host` block, unless the included file also contains `Match`/`Host` directives.

For illustrative purposes:
```
Host example.com
  Include file1 file2  # the directives in file1 & file2 only apply to `example.com`
```

Do you think it would be possible to add these features? I realize this blows up the scope a little bit, but I feel that if we add `Include` support, we should make it work as `Include` is supposed to work.

Reply:
@mislav I agree with you, if we want to add support for the `Include` directive, we should fully support it. I'll take a look at the ssh_config spec and implement a solution for it.